src/peft/tuners/lora/eva.py

```diff
@@ -0,0 +1,505 @@
+        else:
+            fn = prepare_activations_fn
+        hook = SVDHook(name, max_components, peft_config.eva_config.tau, fn)
+        module = model.get_submodule(name)
+        handle = module.register_forward_hook(hook)
+        hooks[name] = (hook, handle)  # adding the old handle here so we dont get errors in the first forward pass
+    layer_hook_map = {**dict(zip(hooks.keys(), hooks.keys())), **equal_inputs_map}
+
+    # start svd calculation
+    if show_progress_bar:
+        pbar = tqdm(iter(cycle(dataloader)), position=0, leave=False)
+    else:
+        pbar = iter(cycle(dataloader))
+    convergence_dict = {k: False for k in hooks.keys()}
+    rank_dist = {k: max_components for k in layer_hook_map.keys()}
+    for inputs in pbar:
+        indices = None
+        if get_indices_fn is not None:
+            indices = get_indices_fn(inputs, peft_config)
+
+        for name in list(hooks.keys()):
+            hook, handle = hooks.get(name)
+            # check if all components that are needed for the rank distribution have converged
+            converged = torch.all(hook.converged[: rank_dist[name]])
+            # if a layer has switched from not converged to converged in the current step
+            if (not convergence_dict[name]) and converged and handle:
+                handle.remove()
+                handle = None
+                convergence_dict[name] = True
+                continue
+            # if a layer has switched from converged to not converged in the current step
+            elif convergence_dict[name] and not converged:
+                module = model.get_submodule(name)
+                handle = module.register_forward_hook(hook)
+                convergence_dict[name] = False
+            hook.indices = indices.T.unbind()
+            hooks[name] = (hook, handle)
+
+        if show_progress_bar:
+            layer_converged = list(convergence_dict.values()) + [
+                convergence_dict[v] for v in equal_inputs_map.values()
+            ]
+            pbar.set_description(f"{sum(layer_converged)}/{len(layer_converged)} layers have converged")
+
+        if all(convergence_dict.values()):
+            break
+
+        forward_fn(model, move_inputs_to_device(inputs, device))
+
+        # in case some hooks have to skip the svd calculation because the number of tokens is less than the number of components
+        if not all(hasattr(h[0].svd, "components_") for h in hooks.values()):
+            continue
+
+        rank_dist = _get_rank_distribution(hooks, layer_hook_map, equal_inputs_map, rank_budget, max_components)
+
+    # check all svd hooks have been removed
+    remaining_svd_hooks = {
+        n for n, m in model.named_modules() for v in m._forward_hooks.values() if isinstance(v, SVDHook)
+    }
+    if len(remaining_svd_hooks) > 0:
+        raise ValueError(
+            f"All SVD hooks should have been removed but the following layers still have active hooks: {remaining_svd_hooks}"
```

Comment:
What should a user do in this case? IIUC, this would be a bug and they should report an issue on GH. If that's the case, please add that instruction to the error message.

Reply:
added a message that an issue should be created